alibi_detect/od/pytorch/pca.py

```diff
@@ -0,0 +1,264 @@
+            n_components: int,
+            device: Optional[Union[str, torch.device]] = None
+            ):
+        """PyTorch backend for PCA detector.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace. For linear PCA should have
+            ``1 <= n_components < dim(data)``. For kernel pca should have ``1 <= n_components < len(data)``.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+
+        Raises
+        ------
+        ValueError
+            If `n_components` is less than 1.
+        """
+        super().__init__(device=device)
+        self.n_components = n_components
+
+        if n_components < 1:
+            raise ValueError('n_components must be at least 1')
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Detect if `x` is an outlier.
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` with leading batch dimension.
+
+        Returns
+        -------
+        `torch.Tensor` of ``bool`` values with leading batch dimension.
+
+        Raises
+        ------
+        ThresholdNotInferredException
+            If called before detector has had `infer_threshold` method called.
+        """
+        scores = self.score(x)
+        if not torch.jit.is_scripting():
+            self.check_threshold_inferred()
+        preds = scores > self.threshold
+        return preds.cpu()
+
+    def score(self, x: torch.Tensor) -> torch.Tensor:
+        """Computes the score of `x`
+
+        Parameters
+        ----------
+        x
+            The tensor of instances. First dimension corresponds to batch.
+
+        Returns
+        -------
+        Tensor of scores for each element in `x`.
+
+        Raises
+        ------
+        NotFitException
+            If called before detector has been fit.
+        """
+        self.check_fitted()
+        score = self.compute_score(x)
+        return score.cpu()
+
+    def fit(self, x_ref: torch.Tensor) -> None:
+        """Fits the PCA detector.
+
+        Parameters
+        ----------
+        x_ref
+            The Dataset tensor.
+        """
+        self.pcs = self._fit(x_ref)
+        self._set_fitted()
+
+    def _fit(self, x: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
+
+    def compute_score(self, x: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class LinearPCATorch(PCATorch):
+    def __init__(
+            self,
+            n_components: int,
+            device: Optional[Union[str, torch.device]] = None
+            ):
+        """Linear variant of the PyTorch backend for PCA detector.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace.
```

Comment:
`principle` or `principal`?

Reply:
Yeah your right, it should be `principal`